man/datemath.1
.\Datetools provide a method of manipulating and working dates and times.
.\Copyright (C) 2013-2018 Chris Caron <lead2gold@gmail.com>
.\
.\This file is part of Datetools.  Datetools is free software; you can
.\redistribute it and/or modify it under the terms of the GNU General Public
.\License as published by the Free Software Foundation; either version 2 of
.\the License, or (at your option) any later version.
.\
.\This program is distributed in the hope that it will be useful,
.\but WITHOUT ANY WARRANTY; without even the implied warranty of
.\MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\GNU General Public License for more details.
.\
.\You should have received a copy of the GNU General Public License along with
.\this program; if not, write to the Free Software Foundation, Inc.,
.\51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
.TH datemath 1 "October 19, 2013" "" "Datemath"

.SH NAME
datemath \- provides a way to perform simple math against the current date and time.

.SH SYNOPSIS
.B datemath
.RI [ options ]
.br

.SH DESCRIPTION
The 
.B datemath
program allows you perform simple math calculations against the current date and time.

.SH OPTIONS
Show this help screen.
.B
.IP -s,	--seconds=VAL
Specify the offset (+/-) in seconds.
.B
.IP -n,	--minutes=VAL
Specify the offset (+/-) in minutes.
.B
.IP -o,	--hours=VAL
Specify the offset (+/-) in hours.
.B
.IP -d,	--days=VAL
Specify the offset (+/-) in days.
.B
.IP -m,	--months=VAL
Specify the offset (+/-) in months.
.B
.IP -y,	--years=VAL
Specify the offset (+/-) in years.
.B
.IP -f,	--format=FORMAT
Specify the desired output format. The default is: %Y-%m-%d %H:%M:%S

FORMAT controls the output.  Interpreted sequences are:

%a     locale’s abbreviated weekday name (e.g., Sun)

%A     locale’s full weekday name (e.g., Sunday)

%b     locale’s abbreviated month name (e.g., Jan)

%B     locale’s full month name (e.g., January)

%c     locale’s date and time (e.g., Thu Mar  3 23:05:25 2005)

%C     century; like %Y, except omit last two digits (e.g., 20)

%d     day of month (e.g, 01)

%D     date; same as %m/%d/%y

%e     day of month, space padded; same as %_d

%F     full date; same as %Y-%m-%d

%g     last two digits of year of ISO week number (see %G)

%G     year of ISO week number (see %V); normally useful only with %V

%h     same as %b

%H     hour (00..23)

%I     hour (01..12)

%j     day of year (001..366)

%k     hour ( 0..23)

%l     hour ( 1..12)

%m     month (01..12)

%M     minute (00..59)

%n     a newline

%N     nanoseconds (000000000..999999999)

%p     locale’s equivalent of either AM or PM; blank if not known

%P     like %p, but lower case

%r     locale’s 12-hour clock time (e.g., 11:11:04 PM)

%R     24-hour hour and minute; same as %H:%M

%n     a newline

%s     seconds since 1970-01-01 00:00:00 UTC

%S     second (00..60)

%T     time; same as %H:%M:%S

%u     day of week (1..7); 1 is Monday

%U     week number of year, with Sunday as first day of week (00..53)

%V     ISO week number, with Monday as first day of week (01..53)

%w     day of week (0..6); 0 is Sunday

%W     week number of year, with Monday as first day of week (00..53)

%x     locale’s date representation (e.g., 12/31/99)

%X     locale’s time representation (e.g., 23:13:48)

%y     last two digits of year (00..99)

%Y     year
.B
.SH "EXIT STATUS"
.B
datemath
exits with 0 if the time could be successfully calculated otherwise a non-zero value is returned.

.SH "COPYRIGHT"
Copyright  ©  2013-2018  Free  Software  Foundation,  Inc.   License  GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl.html>.

This is free software: you are free to change and redistribute it.  There is NO  WARRANTY,  to  the extent permitted by law.
.SH "SEE ALSO"
\fBdateblock\fP(1)
.SH AUTHOR
Chris Caron <lead2gold@gmail.com>



ChangeLog
# Change Log
All notable changes to this project will be documented in this file.

## [1.0.3] - 2018-05-31
### Changed
- fixed padding issue with dates (padding with a zero)
- pypi support

## [1.0.0] - 2017-04-21
### Added
- Drift option now supported inline as part of the cron
- using a plus (+) you can denote the drift entry in the cron. For example:
'0 /5 +4' is the same as writing '0 /5 * * * * 4'.  In this example we would
unblock every 5 minutes but drift an additional 4 seconds.  Inline drifting
allows multiple drifts such as '0 /5 +15,45' giving the same benifits as the
other fields do using this option.
- Internet Systems Consortium (ISC) support added (which is effectively
support for the standard cron format in addition to the enhanced one provided
by dateblock.

## [0.8.1] - 2015-08-17
### Changed
- Fixed bug where month was missing in printout formatting

## [0.8.0] - 2013-10-26
### Added
- Initial commit to GitHub


src/datemath.cpp
/*
 Datetools provide a method of manipulating and working dates and times.
 Copyright (C) 2013-2018 Chris Caron <lead2gold@gmail.com>

 This file is part of Datetools.  Datetools is free software; you can
 redistribute it and/or modify it under the terms of the GNU General Public
 License as published by the Free Software Foundation; either version 2 of
 the License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc.,
 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
*/
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <signal.h>
#include <iostream>
#include <sstream>
#include <string>
#include <boost/program_options.hpp>
#include "Date.h"
#include "datemath.h"

namespace po = boost::program_options;
using namespace std;

// locale’s abbreviated weekday name
const char dayAbbreva[7][4] = {"Sun","Mon","Tue","Wed","Thu","Fri","Sat"};
const int Sun2Mon[7] = {7,1,2,3,4,5,6};
const char dayOfWeek[7][10] = {"Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"};
const char monAbbrev[12][4] = {"Jan","Feb","Mar","Apr","May","Jun","Aug","Sep","Oct","Nov","Dec"};
const char monOfYear[12][10] = {"January","February","March","April","May","June","August","September","October","November","December"};

//--------------------------------------------------
//
//  Function: ProgramDetails
//
//------------------------------------------------
void ProgramDetails()
{
   cerr << "Datemath v" << PACKAGE_VERSION  << " "
        << "Copyright (C) 2013 " << AUTHOR << endl << endl
        << "Dateblock comes with ABSOLUTELY NO WARRANTY" << endl
        << "This is free software, and you are welcome to redistribute it" << endl
        << "under certain conditions; see man pages for details." << endl << endl

        << "This tool can be used to preform math using the current date/time." << endl
        << "Syntax: "
        << "datemath [options] [-f DateTimeFormat]" << endl << endl;
}

//--------------------------------------------------
//
//  Function: FindAndReplace
//
//------------------------------------------------
bool FindAndReplace(string &sourceString, const string &findString, const string &replaceString)
{
   string::size_type findStringLoc;
   bool foundAtLeastOne;

   if (findString == "" || findString == replaceString)
   {
      return false;
   }

   foundAtLeastOne = false;
   findStringLoc = 0;

   while ((findStringLoc = sourceString.find(findString, findStringLoc)) != string::npos)
   {
      sourceString.replace(findStringLoc, findString.length(), replaceString);
      findStringLoc += replaceString.length();

      foundAtLeastOne = true;
   }

   return(foundAtLeastOne);
}
//--------------------------------------------------
//
//  Function: FindAndReplace
//
//------------------------------------------------
bool FindAndReplace(string &sourceString, const string &findString, const int &replaceNum, int width , char fill)
{
   string replaceString;
   string::size_type findStringLoc;
   std::string s;
   std::stringstream out;
   out.width(width);
   out.fill(fill);
   out << replaceNum;
   replaceString = out.str();

   bool foundAtLeastOne;

   if (findString == "" || findString == replaceString)
   {
      return false;
   }

   foundAtLeastOne = false;
   findStringLoc = 0;

   while ((findStringLoc = sourceString.find(findString, findStringLoc)) != string::npos)
   {
      sourceString.replace(findStringLoc, findString.length(), replaceString);
      findStringLoc += replaceString.length();

      foundAtLeastOne = true;
   }

   return(foundAtLeastOne);
}

int main(int argc, char **argv)
{
   Date curTime;

   // Input Parameters
   string  strFormat = "%Y-%m-%d %H:%M:%S";
   long    lSecOffset=0;
   long    lMinOffset=0;
   long    lHourOffset=0;
   long    lDayOffset=0;
   long    lMonthOffset=0;
   long    lYearOffset=0;


   // Declare the supported options.
   po::options_description poAllOptions("Options");
   po::variables_map poVariablesMap;
   poAllOptions.add_options()
       ("help,h", "Show this help screen.")
       ("seconds,s", po::value<long>(), "Specify the offset (+/-) in seconds.")
       ("minutes,n", po::value<long>(), "Specify the offset (+/-) in minutes.")
       ("hours,o", po::value<long>(), "Specify the offset (+/-) in hours.")
       ("days,d", po::value<long>(), "Specify the offset (+/-) in days.")
       ("months,m", po::value<long>(), "Specify the offset (+/-) in months.")
       ("years,y", po::value<long>(), "Specify the offset (+/-) in years.")
       ("format,f", po::value<string>(), "Specify the desired output format (see $>man date). The default is: %Y-%m-%d %H:%M:%S")
   ;


   // Parse all the options and ensure the critical ones have been defined.
   // It's up to the application mainline to ensure application-specific
   // arguments have been specified on the command-line.
   try
   {
      po::store(po::parse_command_line(argc, argv, poAllOptions), poVariablesMap);
   }
   catch(exception& e)
   {
      cerr << "Error parsing command line: " << e.what() << endl;
      cerr << "Exiting." << endl;
      return -1;
   }
   po::notify(poVariablesMap);
   if (poVariablesMap.count("help"))
   {
      ProgramDetails();
      cerr << poAllOptions << endl;
      return 0;
   }

   if (! ( (poVariablesMap.count("seconds")) ||
         (poVariablesMap.count("minutes")) ||
         (poVariablesMap.count("hours")) ||
         (poVariablesMap.count("months")) ||
         (poVariablesMap.count("days")) ||
         (poVariablesMap.count("years"))
         )
      )
   {
      ProgramDetails();
      cerr << poAllOptions << endl;
      return 1;
   }

   // Store Seconds
   if ((poVariablesMap.count("seconds")))
   {
      lSecOffset = poVariablesMap["seconds"].as<long>();
   }
   // Store Minutes
   if ((poVariablesMap.count("minutes")))
   {
      lMinOffset = poVariablesMap["minutes"].as<long>();
   }
   // Store Hours
   if ((poVariablesMap.count("hours")))
   {
      lHourOffset = poVariablesMap["hours"].as<long>();
   }
   // Store Days
   if ((poVariablesMap.count("days")))
   {
      lDayOffset = poVariablesMap["days"].as<long>();
   }
   // Store Months
   if ((poVariablesMap.count("months")))
   {
      lMonthOffset = poVariablesMap["months"].as<long>();
   }
   // Store Years
   if ((poVariablesMap.count("years")))
   {
      lYearOffset = poVariablesMap["years"].as<long>();
   }

   // Store Format
   if ((poVariablesMap.count("format")))
   {
      strFormat = poVariablesMap["format"].as<string>();
   }

   curTime.AddYear(lYearOffset);
   curTime.AddMonth(lMonthOffset);
   curTime.AddDOM(lDayOffset);
   curTime.AddHour(lHourOffset);
   curTime.AddMin(lMinOffset);
   curTime.AddSec(lSecOffset);

   printf("%s\n",curTime.Str(strFormat).c_str());
   return 0;
}
